File: src/main/java/fr/raconteur/simpleskinswapper/SimpleSkinSwapperClient.java
package fr.raconteur.simpleskinswapper;

import fr.raconteur.simpleskinswapper.changeskin.StartupSkinSync;
import fr.raconteur.simpleskinswapper.config.SimpleSkinSwapperConfig;
import fr.raconteur.simpleskinswapper.networking.SkinShuffleCompat;
import fr.raconteur.simpleskinswapper.gui.SkinCarouselScreen;
import fr.raconteur.simpleskinswapper.gui.SkinShuffleImporter;
import fr.raconteur.simpleskinswapper.gui.SkinWheelScreen;
import net.fabricmc.api.ClientModInitializer;
import net.fabricmc.fabric.api.client.keybinding.v1.KeyBindingHelper;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayConnectionEvents;
import net.minecraft.client.option.KeyBinding;
import net.minecraft.client.util.InputUtil;
import net.minecraft.util.Identifier;
import org.lwjgl.glfw.GLFW;

public class SimpleSkinSwapperClient implements ClientModInitializer {

    public static KeyBinding openCarouselKey;
    public static KeyBinding openWheelKey;
    /** Accumulated tick delta for animations (incremented each game tick). */
    public static float TOTAL_TICK_DELTA = 0;

    @Override
    public void onInitializeClient() {
        SkinShuffleCompat.init();
        SkinShuffleImporter.importIfNeeded();
        StartupSkinSync.run();
        KeyBinding.Category category = KeyBinding.Category.create(Identifier.of("simpleskinswapper.title"));
        openCarouselKey = KeyBindingHelper.registerKeyBinding(new KeyBinding(
                "key.simpleskinswapper.open_carousel",
                InputUtil.Type.KEYSYM,
                GLFW.GLFW_KEY_UNKNOWN,
                category
        ));

        openWheelKey = KeyBindingHelper.registerKeyBinding(new KeyBinding(
                "key.simpleskinswapper.open_wheel",
                InputUtil.Type.KEYSYM,
                GLFW.GLFW_KEY_UNKNOWN,
                category
        ));

        ClientPlayConnectionEvents.JOIN.register((handler, sender, client) -> {
            if (client.getCurrentServerEntry() != null) {
                SimpleSkinSwapperConfig.get().registerServerIfAbsent(client.getCurrentServerEntry().address);
            }
        });

        net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents.END_CLIENT_TICK.register(client -> {
            TOTAL_TICK_DELTA++;
            if (openCarouselKey.wasPressed()) {
                client.setScreen(new SkinCarouselScreen(client.currentScreen));
            }
            if (openWheelKey.wasPressed()) {
                client.setScreen(new SkinWheelScreen(client.currentScreen));
            }
        });
    }
}

File: src/main/kotlin/fr/raconteur/simpleskinswapper/SimpleSkinSwapperClient.kt
package fr.raconteur.simpleskinswapper

import com.mojang.blaze3d.platform.InputConstants
import fr.raconteur.simpleskinswapper.changeskin.StartupSkinSync
import fr.raconteur.simpleskinswapper.gui.SkinCarouselScreen
import fr.raconteur.simpleskinswapper.gui.SkinShuffleImporter
import fr.raconteur.simpleskinswapper.gui.SkinWheelScreen
import fr.raconteur.simpleskinswapper.networking.SkinShuffleCompat
import net.fabricmc.api.ClientModInitializer
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents
import net.fabricmc.fabric.api.client.keymapping.v1.KeyMappingHelper
import net.minecraft.client.KeyMapping
import net.minecraft.resources.Identifier

class SimpleSkinSwapperClient : ClientModInitializer {

    override fun onInitializeClient() {
        SkinShuffleCompat.init()
        SkinShuffleImporter.importIfNeeded()
        StartupSkinSync.run()
        val category = KeyMapping.Category.register(Identifier.fromNamespaceAndPath("simpleskinswapper", "title"))
        //? if >=26.3 {
        /*val unboundKeyType = InputConstants.Type.KEYBOARD
        *///?} else {
        val unboundKeyType = InputConstants.Type.KEYSYM
        //?}
        openCarouselKey = KeyMappingHelper.registerKeyMapping(
            KeyMapping(
                "key.simpleskinswapper.open_carousel",
                unboundKeyType,
                InputConstants.UNKNOWN.value,
                category
            )
        )

        openWheelKey = KeyMappingHelper.registerKeyMapping(
            KeyMapping(
                "key.simpleskinswapper.open_wheel",
                unboundKeyType,
                InputConstants.UNKNOWN.value,
                category
            )
        )

        ClientTickEvents.END_CLIENT_TICK.register { client ->
            TOTAL_TICK_DELTA++
            // TEMP-DEBUG: auto-open the YACL config screen for automated GUI testing
            //? if >=26.2 {
            if (debugAutoOpenConfig && client.gui.screen() is net.minecraft.client.gui.screens.TitleScreen) {
            //?} else {
            /*if (debugAutoOpenConfig && client.screen is net.minecraft.client.gui.screens.TitleScreen) {
            *///?}
                debugAutoOpenConfig = false
                //? if >=26.2 {
                client.gui.setScreen(fr.raconteur.simpleskinswapper.gui.config.YaclConfigScreen.create(null))
                //?} else {
                /*client.setScreen(fr.raconteur.simpleskinswapper.gui.config.YaclConfigScreen.create(null))
                *///?}
            }
            if (openCarouselKey?.consumeClick() == true) {
                //? if >=26.2 {
                client.gui.setScreen(SkinCarouselScreen(client.gui.screen()))
                //?} else {
                /*client.setScreen(SkinCarouselScreen(client.screen))
                *///?}
            }
            if (openWheelKey?.consumeClick() == true) {
                //? if >=26.2 {
                client.gui.setScreen(SkinWheelScreen(client.gui.screen()))
                //?} else {
                /*client.setScreen(SkinWheelScreen(client.screen))
                *///?}
            }
        }
    }

    companion object {
        @JvmField
        var openCarouselKey: KeyMapping? = null

        @JvmField
        var openWheelKey: KeyMapping? = null

        /** Accumulated tick delta for animations (incremented each game tick). */
        @JvmField
        var TOTAL_TICK_DELTA = 0f

        // TEMP-DEBUG: set by -Dsss.debugConfig=1 for automated GUI testing
        @JvmField
        var debugAutoOpenConfig: Boolean = System.getProperty("sss.debugConfig") != null
    }
}
